icechunk-python/python/icechunk/session.py

```diff
@@ -233,3 +226,53 @@ def rebase(self, solver: ConflictSolver) -> None:
             When a conflict is detected and the solver fails to resolve it.
         """
         self._session.rebase(solver)
+
+    def fork(self) -> "ForkSession":
+        if self.has_uncommitted_changes:
+            raise ValueError(
+                "Cannot fork a Session with uncommitted changes. "
+                "Make a commit, create a new Session, and then fork that to execute distributed writes."
+            )
+        self._poisoned = True
+        return ForkSession(self._session)
+
+
+class ForkSession(Session):
+    def __getstate__(self) -> object:
+        state = {"_session": self._session.as_bytes()}
+        return state
+
+    def __setstate__(self, state: object) -> None:
+        if not isinstance(state, dict):
+            raise ValueError("Invalid state")
+        self._session = PySession.from_bytes(state["_session"])
+
+    def to_session(self) -> Session:
+        return Session(self._session)
+
+    def merge(self, other: Self) -> None:
+        assert isinstance(other, ForkSession)
```

Comment:
I think this a user error, `ValueError` or `TypeError` as you do with `commit`, with clear message could be better than assert